### ISSUE TITLE
False positive occuring with `cognitive_complexity` on macros

### ISSUE THREAD

**Issue by edward-shen** [rust-lang/rust-clippy#7905]
Lint name: `cognitive_complexity`

I tried this code ([playground](https://gist.github.com/4a43f83b3014f4a1580cee10cbd07db0)):

```rust
#![warn(clippy::cognitive_complexity)]

use tracing::info; // 0.1.26

fn main() {
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
}
```

I expected to see this happen: To not get a cognitive complexity lint.

Instead, this happened: I receive the following lint:

```
warning: the function has a cognitive complexity of (27/25)
 --> src/main.rs:5:4
  |
5 | fn main() {
  |    ^^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::cognitive_complexity)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: you could split it up into multiple smaller functions
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.58.0-nightly (efd048394 2021-10-20)
binary: rustc
commit-hash: efd0483949496b067cd5f7569d1b28cd3d5d3c72
commit-date: 2021-10-20
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

Looks like the lint is expanding the macros out before evaluating cognitive complexity which is causing this to misfire. Suggestions would be to not expand macros when determining cognitive complexity or treating externally defined macros as a normal function call when determining cognitive complexity.

**Comment by edward-shen** [comment]
Example no longer repros. Closing.

**Comment by sivizius** [comment]
Exactly this code copy-pasted into my code triggers `error: the function has a cognitive complexity of (92/25)`.

Versions:

```shell
$ rustc -Vv                                                                           
rustc 1.88.0 (6b00bc388 2025-06-23) (built from a source tarball)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-unknown-linux-gnu
release: 1.88.0
LLVM version: 20.1.6
$ cargo clippy --version --verbose   
clippy 0.1.88
```

But not on the rust playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=ebeb45881388e5af434464e181b0799b

It didn’t triggered in 1.87.0. In fact, it even ignored a lot of very complex functions, afaics those with tracing-macros. Each `info!` seems to increase the complexity by 7, the expanded coded both by the playground and locally expanded looks the same.

I will investigate further why it triggered in my code but not in the playground.
